File: timetaken.go
package main
import "os"
import "fmt"
import "math"
import "bufio"
import "strconv"
import "strings"

func timeTaken(word_dict map[int][2]int, input string) int {
    var timeTakenToType float64
    var current_row, current_col float64
    var first rune
    for _, c := range input {
        first = c
        break
    }
    if rowcol, ok := word_dict[int(first)]; ok {
        current_row, current_col = float64(rowcol[0]), float64(rowcol[1])
    } else {
        return -1
    }
    for _, char := range input[1:] {
        if rowcol, ok := word_dict[int(char)]; ok {
            row, col := float64(rowcol[0]), float64(rowcol[1])
            timeTakenToType += math.Abs(current_row - row) + math.Abs(current_col - col)
        } else {return -1}
    }
    return int(timeTakenToType)
}
func main() {
    word_dict := make(map[int][2]int)
    scanner := bufio.NewScanner(os.Stdin)
    scanner.Scan()
    row_col := scanner.Text()
    var n  int
    for index, value := range strings.Split(row_col, " ") {
        int_value, _ := strconv.Atoi(value)
        if index == 0 {
            n = int_value
        }
    }
    for i := 1; i <= n; i++ {
        scanner.Scan()
        for pos, char := range scanner.Text() {
            rowcol := [2]int{i, pos}
            word_dict[int(char)] = rowcol
        }
    }
    scanner.Scan()

    fmt.Printf("%v\n", timeTaken(word_dict, scanner.Text()))
}

File: time_taken.py
'''
# Sample code to perform I/O:

name = input()                  # Reading input from STDIN
print('Hi, %s.' % name)         # Writing output to STDOUT

# Warning: Printing unwanted or ill-formatted data to output will cause the test cases to fail
'''

# Write your code here

def time_taken(word_dict, string):
    row, col = word_dict.get(string[0], [-1, -1])
    if row == -1:
        print(string[0], "string[0]")
        return -1
    current_row = row
    current_col = col
    typing_time = 0
    for index in range(1, len(string)):
        row, col = word_dict.get(string[index], [-1, -1])
        if row == -1:
            return -1
        typing_time += abs(current_row - row) + abs(current_col - col)
        current_row = row
        current_col = col
    return typing_time

if __name__ == "__main__":
    word_dict = dict()
    row_col = input()
    row, col = row_col.split(" ")
    for i in range(int(row)):
        keyboard_in = input()
        for index, char in enumerate(keyboard_in):
            word_dict[char] = [i, index]
    string = input()
    print(time_taken(word_dict, string))